source/developers-guide/plugin-extension-by-plugin/index.md

```diff
@@ -0,0 +1,684 @@
+     *  
+     * @param InstallContext $context
+     */
+    public function install(InstallContext $context)
+    {
+        $installer = new Installer($this->container);
+        $installer->install();
+    }
+}
+```
+
+### SwagExtendCustomProducts/Installer.php
+To keep track of the code, swap out the installation code in this extra file. 
+In this example it checks if the plugin "Custom Products (v2)" is installed, but you can add other code here if necessary for your installation.
+
+```php
+<?php
+
+namespace SwagExtendCustomProducts\Setup;
+
+use RuntimeException;
+use SwagExtendCustomProducts\Components\Verifier;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class Installer
+{
+    /** @var ContainerInterface */
+    private $container;
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * @return bool
+     */
+    public function install()
+    {
+        // Checks if the plugin Custom Products (v2) is installed.
+        $verifier = new Verifier($this->container->get('dbal_connection'));
+        if(!$verifier->isDependingPluginInstalled()) {
+            throw new RuntimeException('The Plugin "Custom Products (v2)" is required.');
+        }
+
+        return true;
+    }
+}
+```
+
+## Decorate the WhiteListService
+To add new items into the white list we must to decorate the service. 
+That means, call the original service and create a own service. This contains the original service, implements the same interface and extend the functions. 
+At end set the service to the container instead of the original. 
+
+ - To decorate the WhiteListService of the Custom Product (v2) plugin we need the event name that invokes the Service. _custom_products.file_upload.file_type_whitelist_
+ - In combination with the prefix _Enlight_Bootstrap_AfterInitResource_ we create the full event name: **_Enlight_Bootstrap_AfterInitResource_custom_products.file_upload.file_type_whitelist_**
+ 
+ For more information about Shopware events read the <a href="{{ site.url }}/developers-guide/event-guide/">Event guide</a>
+ 
+The Subscriber **_SwagExtendCustomProducts/Subscriber/FileTypeDecorator.php_** 
+
+```php
+<?php
+
+namespace SwagExtendCustomProducts\Subscriber;
+
+use Enlight\Event\SubscriberInterface;
+use ShopwarePlugins\SwagCustomProducts\Components\FileUpload\FileTypeWhitelist;
+use SwagExtendCustomProducts\Decorators\FileTypeWhiteListDecorator;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class FileTypeDecorator implements SubscriberInterface
+{
+    /** @var ContainerInterface */
+    private $container;
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            'Enlight_Bootstrap_AfterInitResource_custom_products.file_upload.file_type_whitelist' => 'decorateFileTypeWhiteList'
```

Comment:
Why are you using the old way here? If you are already using the new plugin system please get rid of the subscriber alltogether and use the services.xml file to decorate an existing service. 

```
<service id="swag_extend_custom_products.file_upload.file_type_whitelist"
         decorates="custom_products.file_upload.file_type_whitelist"
         class="FileTypeWhiteListDecorator">
    <argument type="service" id="swag_extend_custom_products.file_upload.file_type_whitelist.inner"/>
</service>
```